<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [132]:
import pandas as pd
import psycopg2

In [133]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [134]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

0. Выведем структуру БД для понимания, какие таблицы присутствуют

In [135]:
query_0 = f'''
select *
from information_schema.tables t
''' 
total_df = pd.read_sql_query(sql=query_0, con=connection)
display(total_df)

# Для удобство создадим файл xlsx в сторонней папке
total_df.to_excel(excel_writer='Not_important/total_sql_base.xlsx', sheet_name='total_base')

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1121355407.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_df = pd.read_sql_query(sql=query_0, con=connection)


table_catalog        table_schema              table_name  table_type  \
0     project_sql              public                    test  BASE TABLE   
1     project_sql              public                  t_tab1  BASE TABLE   
2     project_sql              public                  t_tab2  BASE TABLE   
3     project_sql              public              operations  BASE TABLE   
4     project_sql              public                 table_1  BASE TABLE   
..            ...                 ...                     ...         ...   
180   project_sql  information_schema   foreign_data_wrappers        VIEW   
181   project_sql  information_schema  foreign_server_options        VIEW   
182   project_sql  information_schema         foreign_servers        VIEW   
183   project_sql  information_schema   foreign_table_options        VIEW   
184   project_sql  information_schema          foreign_tables        VIEW   

    self_referencing_column_name reference_generation  \
0                           None                 None   
1                           None                 None   
2                           None                 None   
3                           None                 None   
4                           None                 None   
..                           ...                  ...   
180                         None                 None   
181                         None                 None   
182                         None                 None   
183                         None                 None   
184                         None                 None   

    user_defined_type_catalog user_defined_type_schema user_defined_type_name  \
0                        None                     None                   None   
1                        None                     None                   None   
2                        None                     None                   None   
3                        None                     None                   None   
4                        None                     None                   None   
..                        ...                      ...                    ...   
180                      None                     None                   None   
181                      None                     None                   None   
182                      None                     None                   None   
183                      None                     None                   None   
184                      None                     None                   None   

    is_insertable_into is_typed commit_action  
0                  YES       NO          None  
1                  YES       NO          None  
2                  YES       NO          None  
3                  YES       NO          None  
4                  YES       NO          None  
..                 ...      ...           ...  
180                 NO       NO          None  
181                 NO       NO          None  
182                 NO       NO          None  
183                 NO       NO          None  
184                 NO       NO          None  

[185 rows x 12 columns]

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [136]:
# текст запроса
# Посчитаем количество вакансий в базе данных
# Запрос в sql

query_3_1 = f'''
select count(*) 
from vacancies
'''

In [137]:
# Вывод расчетов 

total_vacancies = pd.read_sql_query(sql=query_3_1, con=connection)
print(f'Количество вакансий в базе данных - {total_vacancies.iloc[0][0]}')
print()

Количество вакансий в базе данных - 49197



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1825579212.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_vacancies = pd.read_sql_query(sql=query_3_1, con=connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [138]:
# текст запроса
# Посчитаем количество работодателей.
# Запрос в sql

query_3_2 = f'''
select count(*)
from employers
'''

In [139]:
# результат запроса
# Выводим результат на экран

count_employers = pd.read_sql_query(sql=query_3_2, con=connection)
print(f'Количество работодателей в базе данных - {count_employers.iloc[0][0]}')
print()

Количество работодателей в базе данных - 23501



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3344910275.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers = pd.read_sql_query(sql=query_3_2, con=connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [140]:
# текст запроса
# Посчитаем количество регионов.
# Запрос в sql

query_3_3 = f''' 
select count(*)
from areas
'''

In [141]:
# результат запроса
# Выводим результат на экран

total_regions = pd.read_sql_query(sql=query_3_3, con=connection)
print(f'Количество регионов - {total_regions.iloc[0][0]}')
print()

Количество регионов - 1362



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\2516431766.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_regions = pd.read_sql_query(sql=query_3_3, con=connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [142]:
# текст запроса
# Посчитаем количество сфер деятельности.
# Запрос в sql

query_3_4 = f''' 
select count(*)
from industries
''' 

In [143]:
# результат запроса
# Выводим результат на экран

total_industries = pd.read_sql_query(sql=query_3_4, con=connection)
print(f'Количество представленных сфер деятельности - {total_industries.iloc[0][0]}')
print()

Количество представленных сфер деятельности - 294



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3427149412.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  total_industries = pd.read_sql_query(sql=query_3_4, con=connection)


***

**Вывод 1:** 23,5 тысячи работодателей предоставляют 49 тысяч вакансий (в среднем по 2 вакансии на работодателя). Возможно в базе не представлены крупные компании с 100+ вакансиями или много работодателей выставляют только 1 вакансию или не выставляют вакансии вовсе.

**Вывод 2:** Все участники рынка рабочей силы сосредоточены в 1362 регионах. Это значит, что вакансии приходят из абсолютно разных регионов со всей России и ближайшего зарубежья.

**Вывод 3** Количество представленых сфер крайне мало для такой выборки (в среднем 167 вакансий на отрасль). Это модет говорить о специализации базы данных. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [144]:
# текст запроса
# Напишем sql-запрос для определения количества вакансий в регионах
# Сразу применем лимитирующее условие из задания 4.1 (пятерка лидеров)

query_4_1 = f''' 
select
    ar.name "Регион",
    count(v.*) "Количество вакансий"
from areas as ar
left join vacancies as v on ar.id=v.area_id
group by Регион
order by 2 desc
limit 5
'''

In [145]:
# результат запроса
# Выведем результат в виде таблицы в excel

area_vacancies = pd.read_sql_query(sql=query_4_1, con=connection)
display(area_vacancies)
print()

# Ответ на вопрос
print(f"Список регионов с максимальным количеством вакансий: {list(area_vacancies['Регион'])}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3089418932.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  area_vacancies = pd.read_sql_query(sql=query_4_1, con=connection)


Регион  Количество вакансий
0           Москва                 5333
1  Санкт-Петербург                 2851
2            Минск                 2112
3      Новосибирск                 2006
4           Алматы                 1892


Список регионов с максимальным количеством вакансий: ['Москва', 'Санкт-Петербург', 'Минск', 'Новосибирск', 'Алматы']


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [146]:
# текст запроса
# Напишем запрос в sql

query_4_2 = f''' 
select count(*)
from vacancies
where (salary_from is not null) or (salary_to is not null)
'''

In [147]:
# результат запроса
# Выведем на экран ответ

salary_completed = pd.read_sql_query(sql=query_4_2, con=connection)

# Ответ на вопрос
print(f'Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой - {salary_completed.iloc[0][0]}')
print()

Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой - 24073



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\107331342.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salary_completed = pd.read_sql_query(sql=query_4_2, con=connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [148]:
# текст запроса
# Запрос в sql

query_4_3 = f''' 
select 
    round(avg(v.salary_from)) "Нижняя граница",
    round(avg(v.salary_to)) "Верхняя граница" 
from vacancies as v
'''  

In [149]:
# результат запроса
# Веведем результат на экран

mean_salary_boarders = pd.read_sql_query(sql=query_4_3, con=connection)
display(mean_salary_boarders)
print()

# Ответы на вопросы
print(f"Среднее значение нижнего порога зарплаты в вакансиях - {int(mean_salary_boarders['Нижняя граница'].loc[0])}")
print()
print(f"Среднее значение верхнего порога зарплаты в вакансиях - {int(mean_salary_boarders['Верхняя граница'].loc[0])}")


C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3159638449.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  mean_salary_boarders = pd.read_sql_query(sql=query_4_3, con=connection)


Нижняя граница  Верхняя граница
0         71065.0         110537.0


Среднее значение нижнего порога зарплаты в вакансиях - 71065

Среднее значение верхнего порога зарплаты в вакансиях - 110537


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [150]:
# текст запроса
# Напишем sql-запрос для выведения количества вакансий для каждого сочетания типа рабочего графика и типа трудоустройства

query_4_4 = f''' 
select 
    v.schedule График,
    v.employment "Тип трудоустройства",
    count(v.*)
from vacancies as v
group by 1, 2
order by 3 desc
'''

In [151]:
# результат запроса
# Выведем результат запроса

schedule_employment_number = pd.read_sql_query(sql=query_4_4, con=connection)
display(schedule_employment_number)
print()

# Ответ на вопрос
print(f"На втором месте по популярности находится пара {schedule_employment_number['График'][1]} - {schedule_employment_number['Тип трудоустройства'][1]}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1739767337.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schedule_employment_number = pd.read_sql_query(sql=query_4_4, con=connection)


График  Тип трудоустройства  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1


На втором месте по популярности находится пара Удаленная работа - Полная занятость


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [152]:
# текст запроса
# Напишем запрос в sql

query_4_5 = f''' 
select  
    v.experience "Требуемый опыт",
    count(v.*) "Количество вакансий"
from vacancies v
group by v.experience
order by  count(v.*) asc
'''

In [153]:
# результат запроса

experience_vacancies = pd.read_sql_query(sql=query_4_5, con=connection)
display(experience_vacancies)
print()

# Ответ на вопрос
print(f"Опыт работы в порядке возрастания количества вакансий - {tuple(experience_vacancies['Требуемый опыт'])}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1456543215.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience_vacancies = pd.read_sql_query(sql=query_4_5, con=connection)


Требуемый опыт  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152


Опыт работы в порядке возрастания количества вакансий - ('Более 6 лет', 'Нет опыта', 'От 3 до 6 лет', 'От 1 года до 3 лет')


***

**Вывод 1:** Больше всего вакансий размещается в крупных городах (преимущественно столицах): Москва, Санкт-Петербург, Минск, Новосибирск, Алматы.

**Вывод 2:** Зарплата указана только в половине вакансий (в 24 073 из 49 197).

**Вывод 3:** В среднем нижняя граница зарплат по вакансиям составляет 71 тыс. рублей, верхняя- 110,5 тыс. рублей.

**Вывод 4:** Большинство вакансий (72%) предусматривают полный день и полную занятость. Затем идут варианты графиков, но сохраняется режим полной занятости. Меньше вакансий предусматривают иные варианты работы.

**Вывод 5:** Наиболее востребованы специалисты с опытом работы от 1 до 3 лет (53%) и от 3 до 6 лет (29%).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [154]:
# текст запроса
# Напишем запрос в sql

query_5_1 = f''' 
(select 
    emp.name "Работодатель",
    count(v.*) "Количество вакансий"
from employers as emp
left join vacancies as v on emp.id=v.employer_id
group by Работодатель
order by 2 desc
limit 1)

union

(select 
    emp.name "Работодатель",
    count(v.*) "Количество вакансий"
from employers as emp
left join vacancies as v on emp.id=v.employer_id
group by Работодатель
order by 2 desc
limit 1 offset 4)

order by 2 desc
''' 

In [155]:
# результат запроса
# Выведем результат

employer_vacancies = pd.read_sql_query(sql=query_5_1, con=connection)
display(employer_vacancies)
print()

# Ответ на вопрос
print(f"На первом и пятом месте в списке работодателей находятся {employer_vacancies['Работодатель'].iloc[0]} и {employer_vacancies['Работодатель'].iloc[1]}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3889031087.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_vacancies = pd.read_sql_query(sql=query_5_1, con=connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1  Газпром нефть                  331


На первом и пятом месте в списке работодателей находятся Яндекс и Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [156]:
# текст запроса
# Напишем запрос в sql

query_5_2 = f''' 
select 
    ar.name Регион,
    count(emp.*) "Количество работодателей",
    count(vac.*) "Количество вакансий" 
from areas as ar
left join employers as emp on emp.area=ar.id
full join vacancies as vac on ar.id=vac.area_id
group by Регион
having count(vac.*)=0
order by 2 desc
limit 1
''' 

In [157]:
# результат запроса
# Выведем получившуюся таблицу на экран

regions_employers_vacs = pd.read_sql_query(sql=query_5_2, con=connection)
display(regions_employers_vacs)
print()

# Ответ на вопрос
print(f"Регион с наибольшим количеством работодателей - {regions_employers_vacs['Регион'].iloc[0]}.")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\2737541066.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  regions_employers_vacs = pd.read_sql_query(sql=query_5_2, con=connection)


Регион  Количество работодателей  Количество вакансий
0  Россия                       410                    0


Регион с наибольшим количеством работодателей - Россия.


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [158]:
# текст запроса
query_5_3 = f''' 
select 
    emp.name Работодатель,
    count(distinct vac.area_id) "Количество регионов" 
from employers as emp
left join vacancies as vac on vac.employer_id=emp.id 
group by Работодатель
order by 2 desc
'''

In [159]:
# результат запроса
employer_regions = pd.read_sql_query(sql=query_5_3, con=connection)
display(employer_regions)

# Ответ на вопрос
print(f"Наибольшее количество вакансий на 1 работодателя - {employer_regions['Количество регионов'].iloc[0]} (у {employer_regions['Работодатель'].iloc[0]})")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3979372856.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_regions = pd.read_sql_query(sql=query_5_3, con=connection)


Работодатель  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
23170                  СДЕЛКА                    0
23171            Alandr Group                    0
23172                     СДК                    0
23173      Lemon Land Lombard                    0
23174                  Махина                    0

[23175 rows x 2 columns]

Наибольшее количество вакансий на 1 работодателя - 181 (у Яндекс)


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [160]:
# текст запроса
# Запрос sql

query_5_4 = f''' 
select 
    count(emp.*)
from employers as emp
left join employers_industries as ei on emp.id=ei.employer_id
where ei.industry_id is null
''' 

In [161]:
# результат запроса
count_employers_industries = pd.read_sql_query(sql=query_5_4, con=connection)

# Ответ на вопрос
print(f'Количество работодателей с неуказанной отраслью - {count_employers_industries.iloc[0][0]}')
print()

Количество работодателей с неуказанной отраслью - 8419



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\2689251978.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers_industries = pd.read_sql_query(sql=query_5_4, con=connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [162]:
# текст запроса
# Запрос sql

query_5_5 = f''' 
select 
    emp.name Работодатель,
    count(distinct ei.industry_id) "Количество сфер деятельности"
from employers as emp
left join employers_industries as ei on emp.id=ei.employer_id
group by Работодатель
having count(distinct ei.industry_id) = 4
order by 1
offset 2 limit 1
''' 

In [163]:
# результат запроса
employer_industries = pd.read_sql_query(sql=query_5_5, con=connection)
display(employer_industries)
print()

# Ответ на вопрос
print(f"Третья в спсике компания с 4 сферами деятельности - {employer_industries['Работодатель'].iloc[0]}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1637652806.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_industries = pd.read_sql_query(sql=query_5_5, con=connection)


Работодатель  Количество сфер деятельности
0         2ГИС                             4


Третья в спсике компания с 4 сферами деятельности - 2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [164]:
# текст запроса
# Запрос в sql

query_5_6 = f''' 
select count(distinct ei.employer_id)
from employers_industries as ei
left join industries as ind on ei.industry_id=ind.id
where ind.name like 'Разработка программного обеспечения'
''' 

In [165]:
# результат запроса

software_development = pd.read_sql_query(sql=query_5_6, con=connection)

# Ответ на вопрос
print(f'Количество работодателей-разработчиков програмного обеспечения - {software_development.iloc[0][0]}')


Количество работодателей-разработчиков програмного обеспечения - 3553


C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3655441434.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  software_development = pd.read_sql_query(sql=query_5_6, con=connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [166]:
# Код для получения списка городов
# Импортируем библиотеку requests
import requests

# Парсим страницу из википедии и переводим таблицы в формат строк
response = requests.get('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
tables = pd.read_html(response.text)

# Сохраняем нужную таблицу в формате DataFrame и выводим ее на экран
millions = tables[1]
display(millions)

№            Город Население, тыс. чел.  (на 1 января 2023 года)  \
0    1           Москва                                        13 098   
1    2  Санкт-Петербург                                         5 598   
2    3      Новосибирск                                         1 635   
3    4     Екатеринбург                                         1 539   
4    5           Казань                                         1 315   
5    6  Нижний Новгород                                         1 213   
6    7       Красноярск                                         1 197   
7    8        Челябинск                                         1 183   
8    9           Самара                                         1 164   
9   10              Уфа                                         1 158   
10  11   Ростов-на-Дону                                         1 136   
11  12        Краснодар                                         1 121   
12  13             Омск                                         1 111   
13  14          Воронеж                                         1 052   
14  15            Пермь                                         1 027   
15  16        Волгоград                                         1 026   

   Население, тыс. чел.  (итоги переписи  2021 года)[5]  \
0                                              13 010     
1                                               5 602     
2                                               1 634     
3                                               1 544     
4                                               1 309     
5                                               1 226     
6                                               1 188     
7                                               1 190     
8                                               1 173     
9                                               1 145     
10                                              1 142     
11                                              1 099     
12                                              1 126     
13                                              1 058     
14                                              1 034     
15                                              1 028     

   Прирост  относительно  переписи-2010  \
0                                13,1 %   
1                                14,8 %   
2                                10,9 %   
3                                14,4 %   
4                                14,4 %   
5                                   NaN   
6                                21,6 %   
7                                 5,3 %   
8                                 0,3 %   
9                                 7,8 %   
10                                4,9 %   
11                               43,8 %   
12                                  NaN   
13                               18,9 %   
14                                4,3 %   
15                                0,7 %   

   Население, тыс. чел. (итоги переписи  2010 года)[6]  
0                                              11 504   
1                                               4 880   
2                                               1 474   
3                                               1 350   
4                                               1 144   
5                                               1 251   
6                                                 977   
7                                               1 130   
8                                               1 169   
9                                               1 062   
10                                              1 089   
11                                                764   
12                                              1 166   
13                                                890   
14                                                991   
15                                              1 021

In [167]:
# текст запроса
# Запрос в sql

query_5_7 = f''' 
(select 
    ar.name "Регион",
    count(vac.*) "Количество вакансий"
from vacancies as vac
full join areas as ar on ar.id=vac.area_id
full join employers as emp on emp.id=vac.employer_id
where (emp.name like 'Яндекс') and (ar.name in {tuple(millions['Город'])})
group by "Регион"
order by 2 desc)

union all

(select 
    'Total' "Регион",
    count(vac.*) "Количество вакансий"
from vacancies as vac
full join employers as emp on emp.id=vac.employer_id
full join areas as ar on ar.id=vac.area_id
where (emp.name like 'Яндекс') and (ar.name in {tuple(millions['Город'])}))
''' 

In [168]:
# результат запроса
# Выводим результат

count_yandex = pd.read_sql_query(sql=query_5_7, con=connection)
display(count_yandex)
print()

# Ответы на вопросы
print(f"Количество строк в выборке - {count_yandex.shape[0]}")
print()
print(f"Результат в строке 'Total' - {count_yandex[count_yandex['Регион'] == 'Total']['Количество вакансий'].iloc[0]}")

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\3746149260.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_yandex = pd.read_sql_query(sql=query_5_7, con=connection)


Регион  Количество вакансий
0             Total                  485
1            Москва                   54
2   Санкт-Петербург                   42
3      Екатеринбург                   39
4   Нижний Новгород                   36
5       Новосибирск                   35
6           Воронеж                   32
7         Краснодар                   30
8            Самара                   26
9               Уфа                   26
10           Казань                   25
11            Пермь                   25
12   Ростов-на-Дону                   25
13        Волгоград                   24
14       Красноярск                   23
15        Челябинск                   22
16             Омск                   21


Количество строк в выборке - 17

Результат в строке 'Total' - 485


***

**Вывод 1:** Самым крупным работодателем в выборке является Яндекс. Он разместил 1 933 вакансии (8% от общего количества вакансий) в 181 регионе. В целом данная ситуация говорит о монополистической конкуренции со стороны компаний, предлагающих условия труда (много компаний, некто не занимает долю больше 10%, отличающиеся продукты-условия работы и отсутствие ярко выраженной возможности влиять на рынок труда со стороны какой-либо компании). 

**Вывод 2:** У 8 419 работодателей не указана сфера деятельности (отрасль). Это может быть связано с неграмотным заполнением анкеты перед публикацией в системе, отсутствием предложенных вариантов или намеренного скрытия со стороны неблагонадежных компаний.

**Вывод 3:** В выборке представлено 3 553 компании-разработчика програмного обеспечения. Это 15% от общего числа вакансий (23 501). С учетом того, что данные вакансии как правило высокооплачиваемые, можно сделать вывод о высоком спросе на данных специалистов.

**Вывод 4:** Яндекс только в 485 вакансиях (25% от всех вакансий компании) ищет специалистов в городах-миллионниках России. Это говорит о том, что данная компания не сконцентрирована в крупных городах, а предлагает вакансии людям по всей стране и в ближайшем зарубежье.

**Вывод 5:** Выявлен регион, где заявлены работодатели, но нет вакансий (Россия). Это значит, что все работодатели в России стараются конкретнее указать предлагаемое кондидату место работы. Это улучшает возможности поиска для тех, кто ищет работу или анализирует представленную выборку.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [169]:
# текст запроса
# Запрос в sql

query_6_1 = f''' 
select count(vac.*)
from vacancies as vac
where (lower(vac.name) like '%data%') or (lower(vac.name) like '%данн%')
''' 

In [170]:
# результат запроса
# Выводим результат запроса
data_vacancies = pd.read_sql_query(sql=query_6_1, con=connection)

# Ответ на вопрос
print(f' Количество вакансий, имеющих отношение к данным - {data_vacancies.iloc[0][0]}')
print()

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\645630502.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data_vacancies = pd.read_sql_query(sql=query_6_1, con=connection)


 Количество вакансий, имеющих отношение к данным - 1771



2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [202]:
# текст запроса
# Самым оптимальным способом решения нескольких задач будет использование условий (значений where) 
# в виде отдельной переменной "ds_filter", фильтрующих итоговую таблицу

ds_filter = f'''
    ((lower(vac.name) like '%data scientist%') or
    (lower(vac.name) like '%data science%') or
    (lower(vac.name) like '%исследователь данных%') or
    (lower(vac.name) like '%machine learning%') or
    (lower(vac.name) like '%машинн%обучен%') or
    (vac.name like '%ML%' and lower(vac.name) not like '%html%'))
    '''

# Запрос-sql для задания (таблица с вакансиями для джуниоров)
query_6_2 = f''' 
select count(*)
from vacancies as vac
where 
    {ds_filter} and
    ((lower(vac.name) like '%junior%') or 
    (lower(vac.experience) like 'нет опыта') or 
    (lower(vac.employment) like 'стажировка'))
''' 

In [203]:
# результат запроса
# Выводим результаты

ds_juniors = pd.read_sql_query(sql=query_6_2, con=connection)

# Выводим ответ
print(f"Количество вакансий для специалистов с уровнем junior - {ds_juniors.iloc[0][0]}")
print()

Количество вакансий для специалистов с уровнем junior - 51



C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1163268499.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_juniors = pd.read_sql_query(sql=query_6_2, con=connection)


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

*Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [204]:
# текст запроса
# Запрос в sql 

query_6_3 = f''' 
select count(*)
from vacancies as vac
where ({ds_filter}) and
(vac.key_skills ilike '%postgres%' or vac.key_skills ilike '%sql%')

''' 

In [205]:
# результат запроса
# Результирующая переменная
sql_postgress = pd.read_sql_query(sql=query_6_3, con=connection)
display(sql_postgress)

# Ответ на вопрос
print(f'Количество вакансий с sql и postgres в навыках - {sql_postgress.iloc[0][0]}')
print()

C:\Users\qwert\AppData\Local\Temp\ipykernel_19216\1549929545.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql_postgress = pd.read_sql_query(sql=query_6_3, con=connection)


count
0    201

Количество вакансий с sql и postgres в навыках - 201



4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [175]:
# текст запроса

In [176]:
# результат запроса

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [177]:
# текст запроса

In [178]:
# результат запроса

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [179]:
# текст запроса

In [180]:
# результат запроса

***

In [181]:
# выводы по предметному анализу

# Общий вывод по проекту

In [182]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования